File: src/part3/checkout.ts
/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Typescript handler API,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/tour-of-restate-typescript
 */

import * as restate from "@restatedev/restate-sdk";
import { v4 as uuid } from "uuid";
import { PaymentClient } from "../auxiliary/payment_client";
import { EmailClient } from "../auxiliary/email_client";

const checkout = async (
  ctx: restate.RpcContext,
  request: { userId: string; tickets: string[] },
) => {
  // Generate idempotency key for the stripe client
  const idempotencyKey = await ctx.sideEffect(async () => uuid());

  // We are a uniform shop where everything costs 40 USD
  const totalPrice = request.tickets.length * 40;

  const paymentClient = PaymentClient.get();

  const success = await paymentClient.call(idempotencyKey, totalPrice);

  const email = EmailClient.get();

  if (success) {
    console.info("Payment successful. Notifying user about shipment.");
    await ctx.sideEffect(async () =>
      email.notifyUserOfPaymentSuccess(request.userId),
    );
  } else {
    console.info("Payment failure. Notifying user about it.");
    await ctx.sideEffect(async () =>
      email.notifyUserOfPaymentFailure(request.userId),
    );
  }

  return success;
};

export const checkoutRouter = restate.router({
  checkout: checkout,
});

export const checkoutApi: restate.ServiceApi<typeof checkoutRouter> = {
  path: "Checkout",
};

File: java/src/main/java/dev/restate/tour/part3/Checkout.java
package dev.restate.tour.part3;

import com.google.protobuf.BoolValue;
import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.sdk.common.TerminalException;
import dev.restate.tour.auxiliary.EmailClient;
import dev.restate.tour.auxiliary.PaymentClient;
import dev.restate.tour.generated.CheckoutRestate;
import dev.restate.tour.generated.Tour.CheckoutFlowRequest;

import java.util.UUID;

public class Checkout extends CheckoutRestate.CheckoutRestateImplBase {

    PaymentClient paymentClient = PaymentClient.get();
    EmailClient emailClient = EmailClient.get();

    @Override
    public BoolValue checkout(RestateContext ctx, CheckoutFlowRequest request) throws TerminalException {
        // Generate idempotency key for the stripe client
        var idempotencyKey = ctx.sideEffect(CoreSerdes.STRING_UTF8, () -> UUID.randomUUID().toString());

        // We are a uniform shop where everything costs 40 USD
        var totalPrice = request.getTicketsList().size() * 40.0;

        boolean success = ctx.sideEffect(CoreSerdes.BOOLEAN, () -> paymentClient.call(idempotencyKey, totalPrice));

        if (success) {
            ctx.sideEffect(()-> emailClient.notifyUserOfPaymentSuccess(request.getUserId()));
        } else {
            ctx.sideEffect(() -> emailClient.notifyUserOfPaymentFailure(request.getUserId()));
        }

        return BoolValue.of(success);
    }
}
